File: examples/typescript/src/primitives/channel/channel-flow.ts
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Channel,
  StepList,
  Timer,
  Wait,
  doubleCodec,
  goTo,
  gracefulComplete,
  jsonCodec,
  rpc,
  stringCodec,
  type Context,
  type Flow,
  type PersistenceSchema,
  type RPCResult,
  type Step,
  type StepDecision,
  type StepOptions,
} from "@superdurable/dex";

const approvalMessages = new Channel("ApprovalMessages", stringCodec);
export const queuedMessages = new Channel("QueuedMessages", stringCodec);
export const prioritizedMessages = new Channel("PrioritizedMessages", stringCodec);

export interface QueuedMessageReference {
  readonly messageId: string;
}

export interface PendingMessage {
  readonly messageId: string;
  readonly value: string;
}

const queuedMessageReferenceCodec = jsonCodec<QueuedMessageReference>();
const pendingMessagesCodec = jsonCodec<readonly PendingMessage[]>();

class ChannelWait implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public getStepType(): string {
    return "ChannelWait";
  }

  public getStepOptions(): StepOptions {
    return { executeLoadChannels: [queuedMessages] };
  }

  public waitFor(_context: Context, input: number): Wait {
    return Wait.anyOf(
      approvalMessages.forOne(),
      Timer.byDuration(input * 1000),
    );
  }

  public execute(context: Context, _input: number): StepDecision {
    const pendingQueuedMessages = queuedMessages.pendingMessages(context);
    if (pendingQueuedMessages.length > 0) {
      queuedMessages.delete(context, pendingQueuedMessages[0]!.messageId);
      return gracefulComplete(pendingQueuedMessages[0]!.value);
    }
    if (context.hasTimerFired()) {
      return gracefulComplete("approval timed out");
    }
    const approvalMessageValues = approvalMessages.results(context);
    return gracefulComplete(approvalMessageValues[0]!);
  }
}

const channelWaitStep = new ChannelWait();

export class ChannelFlow implements Flow<number> {
  public getFlowType(): string {
    return "ChannelFlow";
  }

  public getSteps() {
    return StepList.startStep(channelWaitStep);
  }

  public getPersistenceSchema(): PersistenceSchema {
    return { channels: [approvalMessages, queuedMessages, prioritizedMessages] };
  }

  @rpc()
  public publishApprovalMessage(context: Context): void {
    approvalMessages.publish(context, "approved");
  }

  @rpc({ inputCodec: stringCodec })
  public enqueueChannelMessage(context: Context, value: string): void {
    queuedMessages.publish(context, value);
  }

  @rpc({ loadChannels: [queuedMessages], outputCodec: pendingMessagesCodec })
  public getQueuedMessages(context: Context): RPCResult<readonly PendingMessage[]> {
    return { output: queuedMessages.pendingMessages(context) };
  }

  @rpc({
    isTransactional: true,
    loadChannels: [queuedMessages],
    inputCodec: queuedMessageReferenceCodec,
  })
  public deleteQueuedMessage(context: Context, queuedMessage: QueuedMessageReference): void {
    queuedMessages.delete(context, queuedMessage.messageId);
  }

  @rpc({ loadChannels: [prioritizedMessages], outputCodec: pendingMessagesCodec })
  public getPrioritizedMessages(context: Context): RPCResult<readonly PendingMessage[]> {
    return { output: prioritizedMessages.pendingMessages(context) };
  }

  @rpc({
    isTransactional: true,
    loadChannels: [queuedMessages],
    inputCodec: queuedMessageReferenceCodec,
  })
  public moveQueuedMessageToPrioritizedMessages(
    context: Context,
    queuedMessage: QueuedMessageReference,
  ): void {
    const messageToPrioritize = queuedMessages.findPendingMessage(
      context,
      queuedMessage.messageId,
    );
    queuedMessages.delete(context, queuedMessage.messageId);
    if (messageToPrioritize !== undefined) {
      prioritizedMessages.publish(context, messageToPrioritize.value);
    }
  }
}

export const channelFlow = new ChannelFlow();

File: examples/python/dex_examples/primitives/channel/channel_flow.py
# Copyright (c) 2022-2026 Super Durable, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the applicable language governing permissions and
# limitations under the License.

"""Minimal Channel Flow: publish on an RPC and wait in a Step."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import timedelta

from dex import (
    Channel,
    Context,
    Flow,
    PersistenceSchema,
    RPCResult,
    Step,
    StepDecision,
    StepList,
    StepOptions,
    Timer,
    Wait,
    go_to,
    graceful_complete,
    rpc,
)


@dataclass(frozen=True)
class QueuedMessageReference:
    message_id: str


@dataclass(frozen=True)
class PendingMessage:
    message_id: str
    value: str


class ChannelWaitStep(Step[int]):
    def __init__(
        self,
        approval_messages: Channel[str],
        queued_messages: Channel[str],
    ) -> None:
        self.approval_messages = approval_messages
        self.queued_messages = queued_messages

    def get_step_options(self) -> StepOptions:
        return StepOptions(execute_load_channels=(self.queued_messages,))

    def wait_for(self, context: Context, input: int) -> Wait:
        return Wait.any_of(
            self.approval_messages.for_one(),
            Timer.by_duration(timedelta(seconds=input)),
        )

    def execute(self, context: Context, input: int) -> StepDecision:
        pending_queued_messages = self.queued_messages.pending_messages(context)
        if pending_queued_messages:
            self.queued_messages.delete(context, pending_queued_messages[0].message_id)
            return graceful_complete(pending_queued_messages[0].value)
        if context.has_timer_fired():
            return graceful_complete("approval timed out")
        approval_message_values = self.approval_messages.results(context)
        return graceful_complete(approval_message_values[0])


class ChannelFlow(Flow[int]):
    approval_messages = Channel("ApprovalMessages", str)
    queued_messages = Channel("QueuedMessages", str)
    prioritized_messages = Channel("PrioritizedMessages", str)

    def __init__(self) -> None:
        self.wait_for_approval = ChannelWaitStep(
            self.approval_messages, self.queued_messages
        )

    def get_steps(self) -> StepList[int]:
        return StepList.start_step(self.wait_for_approval)

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(
            self.approval_messages,
            self.queued_messages,
            self.prioritized_messages,
        )

    @rpc
    def publish_approval_message(self, context: Context) -> None:
        self.approval_messages.publish(context, "approved")

    @rpc
    def enqueue_channel_message(self, context: Context, input: str) -> None:
        self.queued_messages.publish(context, input)

    @rpc(load_channels=(queued_messages,))
    def get_queued_messages(
        self, context: Context
    ) -> RPCResult[list[PendingMessage]]:
        return RPCResult(
            [PendingMessage(message.message_id, message.value)
             for message in self.queued_messages.pending_messages(context)]
        )

    @rpc(is_transactional=True, load_channels=(queued_messages,))
    def delete_queued_message(
        self, context: Context, queued_message: QueuedMessageReference
    ) -> None:
        self.queued_messages.delete(context, queued_message.message_id)

    @rpc(load_channels=(prioritized_messages,))
    def get_prioritized_messages(
        self, context: Context
    ) -> RPCResult[list[PendingMessage]]:
        return RPCResult(
            [PendingMessage(message.message_id, message.value)
             for message in self.prioritized_messages.pending_messages(context)]
        )

    @rpc(is_transactional=True, load_channels=(queued_messages,))
    def move_queued_message_to_prioritized_messages(
        self, context: Context, queued_message: QueuedMessageReference
    ) -> None:
        message_to_prioritize = self.queued_messages.find_pending_message(
            context, queued_message.message_id
        )
        self.queued_messages.delete(context, queued_message.message_id)
        if message_to_prioritize is not None:
            self.prioritized_messages.publish(context, message_to_prioritize.value)
